presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java

```diff
@@ -747,4 +749,67 @@ public boolean cancel(boolean mayInterruptIfRunning)
             return false;
         }
     }
+
+    // Helper class for combining rows from multiple pages into a single List.
+    // Not thread safe.
+    // TODO: Replace RuntimeExceptions with better alternatives.
+    private static class RowData
+    {
+        private List<List<Object>> rows = new ArrayList<>();
+        private boolean rowsFinalized;
+        private long dataSize;
+
+        public RowData()
+        {
+        }
+
+        public void addPage(ConnectorSession session, List<Type> types, Page page)
+        {
+            requireNonNull(session, "session is null");
+            requireNonNull(types, "types is null");
+            requireNonNull(page, "page is null");
+            if (rowsFinalized) {
+                throw new RuntimeException("Can't add more pages once finalized");
+            }
+            for (int position = 0; position < page.getPositionCount(); position++) {
+                List<Object> row = new ArrayList<>(page.getChannelCount());
+                int rowSize = 0;
+                for (int channel = 0; channel < page.getChannelCount(); channel++) {
+                    Type type = types.get(channel);
+                    Block block = page.getBlock(channel);
+                    row.add(type.getObjectValue(session, block, position));
+
+                    // TODO: Find a better way of determining size of the type.getObjectValue returned in previous statement.
+                    // Using type.getSlice() has significant issues.
+                    //   (1) getSlice is not implemented for all the types which support getObjectValue.
```

Comment:
`getObjectValue` must be JSON serializable. You could serialize individual objects and measure how much is that. I don't know how to avoid repeated serialization though. @electrum ?